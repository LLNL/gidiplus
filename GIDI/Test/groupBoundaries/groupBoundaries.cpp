/*
# <<BEGIN-copyright>>
# Copyright 2019, Lawrence Livermore National Security, LLC.
# See the top-level COPYRIGHT file for details.
# 
# SPDX-License-Identifier: MIT
# <<END-copyright>>
*/

#include <stdlib.h>
#include <iostream>

#include "GIDI_testUtilities.hpp"

static char const *description = "This program prints the multi-group boundaries for each transportable particle in a GNDS file.";

void main2( int argc, char **argv );
void printVector( std::string const &prefix, std::string const &indent, std::vector<double> const &vector );
/*
=========================================================
*/
int main( int argc, char **argv ) {

    try {
        main2( argc, argv ); }
     catch (std::exception &exception) {
        std::cerr << exception.what( ) << std::endl;
        exit( EXIT_FAILURE ); }
    catch (char const *str) {
        std::cerr << str << std::endl;
        exit( EXIT_FAILURE ); }
    catch (std::string &str) {
        std::cerr << str << std::endl;
        exit( EXIT_FAILURE );
    }

    exit( EXIT_SUCCESS );
}
/*
=========================================================
*/
void main2( int argc, char **argv ) {

    argvOptions argv_options( "groupBoundaries", description );
    ParseTestOptions parseTestOptions( argv_options, argc, argv );

    parseTestOptions.m_askGNDS_File = true;

    parseTestOptions.parse( );

    GIDI::Construction::PhotoMode photo_mode = parseTestOptions.photonMode( GIDI::Construction::PhotoMode::nuclearAndAtomic );
    GIDI::Construction::Settings construction( GIDI::Construction::ParseMode::all, photo_mode );
    PoPI::Database pops;
    GIDI::Protare *protare = parseTestOptions.protare( pops, "../pops.xml", "../all.map", construction, PoPI::IDs::neutron, "O16" );

    std::cout << stripDirectoryBase( protare->fileName( ), "/GIDI/Test/" ) << std::endl;

    GIDI::Styles::TemperatureInfos temperatures = protare->temperatures( );
    for( GIDI::Styles::TemperatureInfos::iterator iter = temperatures.begin( ); iter != temperatures.end( ); ++iter ) {
        GIDI::PhysicalQuantity const &temperature = iter->temperature( );

        std::cout << "label = " << iter->heatedMultiGroup( ) << "  temperature = " << temperature.value( ) << std::endl;
    }

    std::string label( temperatures[0].heatedMultiGroup( ) );
    std::cout << "label = " << label << std::endl;
    GIDI::Transporting::MG settings( protare->projectile( ).ID( ), GIDI::Transporting::Mode::multiGroup, GIDI::Transporting::DelayedNeutrons::on );

    try {
        std::vector<double> groupBoundaries = protare->groupBoundaries( settings, temperatures[0], protare->projectile( ).ID( ) );
        std::string prefix( "group boundaries" );
        printVector( prefix, "    ", groupBoundaries ); }
    catch (char const *str) {
        std::cout << str << std::endl;
        exit( EXIT_FAILURE );
    }

    GIDI::Styles::HeatedMultiGroup const &style = *(protare->styles( ).get<GIDI::Styles::HeatedMultiGroup>( label ));
    std::cout << style.moniker( ) << std::endl;
    GIDI::Styles::MultiGroup const &multiGroupStyle = style.multiGroup( );
    std::cout << multiGroupStyle.moniker( ) << std::endl;
    label = multiGroupStyle.label( );
    std::cout << "label = " << label << std::endl;
    try {
        std::vector<double> groupBoundaries = protare->groupBoundaries( settings, temperatures[0], protare->projectile( ).ID( ) );
        std::string prefix( "group boundaries" );
        printVector( prefix, "    ", groupBoundaries ); }
    catch (char const *str) {
        std::cout << str << std::endl;
        exit( EXIT_FAILURE );
    }

    std::cout << std::endl;
    GIDI::Styles::MultiGroup const &multiGroup = *(protare->multiGroup( temperatures[0].heatedMultiGroup( ) ));
    std::cout << "multiGroup for label " << temperatures[0].heatedMultiGroup( ) << std::endl;
    std::cout << "    maximumLegendreOrder = " << multiGroup.maximumLegendreOrder( ) << std::endl;

    GIDI::Suite const &transportables = multiGroup.transportables( );
    for( std::size_t i1 = 0; i1 < transportables.size( ); ++i1 ) {
        GIDI::Transportable const &transportable = *(transportables.get<GIDI::Transportable const>( i1 ));

        std::cout << "    " << transportable.label( ) << std::endl;
        std::cout << "        pid = " << transportable.pid( ) << std::endl;

        GIDI::Group const &group = transportable.group( );
        std::cout << "        group name = " << group.label( ) << std::endl;

        GIDI::Grid const &grid = group.grid( );
        std::cout << "        grid unit = " << grid.unit( ) << std::endl;
        printVector( "        values", "            ", grid.data( ) );
    }

    delete protare;
}
/*
=========================================================
*/
void printVector( std::string const &prefix, std::string const &indent, std::vector<double> const &vector ) {

    std::cout << prefix << ": size = " << vector.size( ) << std::endl;
    std::cout << indent;
    for( std::string::size_type index = 0; index < vector.size( ); ++index ) std::cout << " " << vector[index];
    std::cout << std::endl;
}
