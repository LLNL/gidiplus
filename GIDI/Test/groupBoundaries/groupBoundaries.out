/.../GIDI/Test/Data/MG_MC/neutrons/n-008_O_016.xml
label = MultiGroup_000  temperature = 2.586e-08
label = MultiGroup_000
group boundaries: size = 88
     1.3068e-09 2.0908e-08 1.3068e-07 3.3453e-07 1.1761e-06 2.0908e-06 5.6578e-06 1.3068e-05 2.0746e-05 5.123e-05 0.00010245 0.00020908 0.00038105 0.00056578 0.00071558 0.0010585 0.0013068 0.0018817 0.0029402 0.0033453 0.0042339 0.0057628 0.007527 0.010245 0.015106 0.020908 0.026462 0.032669 0.03953 0.07002 0.098909 0.13068 0.18195 0.20746 0.2417 0.27097 0.29402 0.33453 0.37765 0.5123 0.63247 0.7527 0.88337 1.0245 1.1761 1.3381 1.5106 1.6936 2.0908 2.3051 2.5299 2.7411 3.0108 3.5335 4.0688 4.396 4.7044 4.9908 5.3525 5.6578 6.0425 6.3666 6.7367 7.1558 7.5479 7.9096 8.3215 8.7867 9.1767 9.6648 10.12 10.585 11.012 11.547 11.993 12.499 13.068 13.542 13.863 14.134 14.407 14.683 15.186 15.754 16.334 16.923 18.134 20
heatedMultiGroup
multiGroup
label = MultiGroup
group boundaries: size = 88
     1.3068e-09 2.0908e-08 1.3068e-07 3.3453e-07 1.1761e-06 2.0908e-06 5.6578e-06 1.3068e-05 2.0746e-05 5.123e-05 0.00010245 0.00020908 0.00038105 0.00056578 0.00071558 0.0010585 0.0013068 0.0018817 0.0029402 0.0033453 0.0042339 0.0057628 0.007527 0.010245 0.015106 0.020908 0.026462 0.032669 0.03953 0.07002 0.098909 0.13068 0.18195 0.20746 0.2417 0.27097 0.29402 0.33453 0.37765 0.5123 0.63247 0.7527 0.88337 1.0245 1.1761 1.3381 1.5106 1.6936 2.0908 2.3051 2.5299 2.7411 3.0108 3.5335 4.0688 4.396 4.7044 4.9908 5.3525 5.6578 6.0425 6.3666 6.7367 7.1558 7.5479 7.9096 8.3215 8.7867 9.1767 9.6648 10.12 10.585 11.012 11.547 11.993 12.499 13.068 13.542 13.863 14.134 14.407 14.683 15.186 15.754 16.334 16.923 18.134 20

multiGroup for label MultiGroup_000
    maximumLegendreOrder = 3
    He4
        pid = He4
        group name = LLNL_gid_71
        grid unit = MeV
        values: size = 64
             1e-10 0.02 0.0227857 0.0259594 0.0295752 0.0336945 0.0383877 0.0437345 0.049826 0.056766 0.0646727 0.0736806 0.0839432 0.0956352 0.108956 0.124132 0.141421 0.161119 0.183561 0.209128 0.238256 0.271442 0.309249 0.352323 0.401397 0.457305 0.521001 0.593568 0.676243 0.770434 0.877744 1 1.09814 1.20591 1.32425 1.45422 1.59693 1.75365 1.92575 2.11474 2.32228 2.55019 2.80046 3.07529 3.3771 3.70852 4.07247 4.47214 4.91103 5.39299 5.92225 6.50345 7.14169 7.84256 8.61222 9.45742 10.3856 11.4048 12.524 13.7531 15.1028 16.585 18.2126 20
    He3
        pid = He3
        group name = LLNL_gid_71
        grid unit = MeV
        values: size = 64
             1e-10 0.02 0.0227857 0.0259594 0.0295752 0.0336945 0.0383877 0.0437345 0.049826 0.056766 0.0646727 0.0736806 0.0839432 0.0956352 0.108956 0.124132 0.141421 0.161119 0.183561 0.209128 0.238256 0.271442 0.309249 0.352323 0.401397 0.457305 0.521001 0.593568 0.676243 0.770434 0.877744 1 1.09814 1.20591 1.32425 1.45422 1.59693 1.75365 1.92575 2.11474 2.32228 2.55019 2.80046 3.07529 3.3771 3.70852 4.07247 4.47214 4.91103 5.39299 5.92225 6.50345 7.14169 7.84256 8.61222 9.45742 10.3856 11.4048 12.524 13.7531 15.1028 16.585 18.2126 20
    H2
        pid = H2
        group name = LLNL_gid_71
        grid unit = MeV
        values: size = 64
             1e-10 0.02 0.0227857 0.0259594 0.0295752 0.0336945 0.0383877 0.0437345 0.049826 0.056766 0.0646727 0.0736806 0.0839432 0.0956352 0.108956 0.124132 0.141421 0.161119 0.183561 0.209128 0.238256 0.271442 0.309249 0.352323 0.401397 0.457305 0.521001 0.593568 0.676243 0.770434 0.877744 1 1.09814 1.20591 1.32425 1.45422 1.59693 1.75365 1.92575 2.11474 2.32228 2.55019 2.80046 3.07529 3.3771 3.70852 4.07247 4.47214 4.91103 5.39299 5.92225 6.50345 7.14169 7.84256 8.61222 9.45742 10.3856 11.4048 12.524 13.7531 15.1028 16.585 18.2126 20
    H3
        pid = H3
        group name = LLNL_gid_71
        grid unit = MeV
        values: size = 64
             1e-10 0.02 0.0227857 0.0259594 0.0295752 0.0336945 0.0383877 0.0437345 0.049826 0.056766 0.0646727 0.0736806 0.0839432 0.0956352 0.108956 0.124132 0.141421 0.161119 0.183561 0.209128 0.238256 0.271442 0.309249 0.352323 0.401397 0.457305 0.521001 0.593568 0.676243 0.770434 0.877744 1 1.09814 1.20591 1.32425 1.45422 1.59693 1.75365 1.92575 2.11474 2.32228 2.55019 2.80046 3.07529 3.3771 3.70852 4.07247 4.47214 4.91103 5.39299 5.92225 6.50345 7.14169 7.84256 8.61222 9.45742 10.3856 11.4048 12.524 13.7531 15.1028 16.585 18.2126 20
    H1
        pid = H1
        group name = LLNL_gid_71
        grid unit = MeV
        values: size = 64
             1e-10 0.02 0.0227857 0.0259594 0.0295752 0.0336945 0.0383877 0.0437345 0.049826 0.056766 0.0646727 0.0736806 0.0839432 0.0956352 0.108956 0.124132 0.141421 0.161119 0.183561 0.209128 0.238256 0.271442 0.309249 0.352323 0.401397 0.457305 0.521001 0.593568 0.676243 0.770434 0.877744 1 1.09814 1.20591 1.32425 1.45422 1.59693 1.75365 1.92575 2.11474 2.32228 2.55019 2.80046 3.07529 3.3771 3.70852 4.07247 4.47214 4.91103 5.39299 5.92225 6.50345 7.14169 7.84256 8.61222 9.45742 10.3856 11.4048 12.524 13.7531 15.1028 16.585 18.2126 20
    photon
        pid = photon
        group name = LLNL_gid_70
        grid unit = MeV
        values: size = 41
             0.1978 0.2176 0.2394 0.2633 0.2897 0.3186 0.3505 0.3855 0.4241 0.4665 0.5132 0.5645 0.6209 0.683 0.7513 0.8264 0.9091 1 1.125 1.266 1.424 1.602 1.802 2.027 2.281 2.566 2.887 3.247 3.653 4.11 4.624 5.202 5.852 6.583 7.406 8.332 9.373 10.55 11.86 13.35 20
    n
        pid = n
        group name = LLNL_gid_4
        grid unit = MeV
        values: size = 88
             1.3068e-09 2.0908e-08 1.3068e-07 3.3453e-07 1.1761e-06 2.0908e-06 5.6578e-06 1.3068e-05 2.0746e-05 5.123e-05 0.00010245 0.00020908 0.00038105 0.00056578 0.00071558 0.0010585 0.0013068 0.0018817 0.0029402 0.0033453 0.0042339 0.0057628 0.007527 0.010245 0.015106 0.020908 0.026462 0.032669 0.03953 0.07002 0.098909 0.13068 0.18195 0.20746 0.2417 0.27097 0.29402 0.33453 0.37765 0.5123 0.63247 0.7527 0.88337 1.0245 1.1761 1.3381 1.5106 1.6936 2.0908 2.3051 2.5299 2.7411 3.0108 3.5335 4.0688 4.396 4.7044 4.9908 5.3525 5.6578 6.0425 6.3666 6.7367 7.1558 7.5479 7.9096 8.3215 8.7867 9.1767 9.6648 10.12 10.585 11.012 11.547 11.993 12.499 13.068 13.542 13.863 14.134 14.407 14.683 15.186 15.754 16.334 16.923 18.134 20
